docs/adr/0008-freeing-objects-and-resources.md

```diff
@@ -0,0 +1,109 @@
+  `close()` method is called.  Alternatively, handle this by panicking or throwing an error.
+* Use the `Cleaner` API on Kotlin to automatically free UniFFI pointers for Kotlin objects.
+  Note that this combines with the previous bullet.  `close()` releases the resource, the Cleaner
+  action frees the heap allocation.
+* Initially, don't implement `Autoclosable` anymore.  If users want this feature, we could
+  consider supporting `Closable` as a standard trait, like we do with `Debug`, `Display`, etc.
+  In this case, `Closable` wouldn't change the Rust code, instead it would implement
+  `Autoclosable` on Kotlin and maybe implement interfaces for other languages as well.
+* Consider adding a Kotlin configuration option to generate a `destroy()` method to free the Rust
+  pointer instead of relying on the Cleaner.  This option would be for users who need to support
+  older Java versions where Cleaner is not available and users who don't trust that Cleaner is
+  implemented correctly.
+
+### [Option 2] Resource and Object as separate types
+
+* Require UDL users to label their interfaces as either a `Resource` or `Object` and proc-macro
+  users to derive either `Resource` or `Object`.
+* Each type would have slightly different handling.  For example on Kotlin `Resource` would work
+  like an `Object` does today and `Object` would work like the `Object` from [1].
+* For other languages, we could either introduce similar distinctions or make both types behave
+  the same.
+
+### [Option 2] Make Object do both
+
+* Kotlin objects keep their current behavior.  They still have a `destroy` method and also
+  implement `AutoCloseable`.
+* In addition, Kotlin objects also use the `Cleaner` API to ensure they are destroyed if the
+  user doesn't manually call it.
+
+## Pros and Cons of the Options
+
+### [Option 1] User-defined resource close method
+
+* Good, because it's the only option that supports callback interfaces (and trait interfaces in the
+  future).  If a foreign object is passed into Rust and that object holds a resource, Rust needs to
+  have a method it can call to free the resource.  We cannot generate an implementation of that
+  method like we currently do for objects, since there's not one way to close resources in foreign
+  languages (unlike Rust, where dropping an object releases any resources).
+* Good, because it's the most explicit.
+* Good, because gives the user control over their object namespace.  UniFFI doesn't define any methods for them.
+* Bad, because it's not idiomatic to call `close()` on Rust or other RAII-based languages.
+* Bad, because it requires Rust and other RAII-based language implementations to manually check if
+  the resource object has been closed or not.
+* Bad, because it relies on the `Cleaner` Java API, which is not available on very old versions of
+  Android.
+
+### [Option 2] Resource and Object as separate types
```

Comment:
I'd say it also has the downside of adding more complexity. The other two options don't really introduce any new concepts in the public API from my POV, this one does.